JPEG XL support

surma opened the pull request:
This PR lands support for JPEG XL (or JXL).

The encoder currently crashes for some speed settings, but not for others. Not quite sure what’s happening there. 

Options UI will need polish, but the encoder options are bit mysterious to me as of yet. 
@jakearchibald said he’ll do the options UI once it’s good to go.

Closes #721

surma committed:
Compile JXL decoder

surma committed:
Add test to JXL decoder

surma committed:
Integrate JXL decoder

surma committed:
Compile JXL encoder

surma committed:
Integrate encoder

surma committed:
Fix JXL encoder

surma committed:
Exposed speed slider, but it's weird

surma commented:
Related #721

surma committed:
Update JXL library and build system

surma committed:
Update for encoder API changes

surma commented:
Using the Squoosh logo as an example image, I see the following behavior:

- Speed=7: Success
- Speed=6: Success
- Speed=5: Success
- Speed=4: Success
- Speed=3: Success
- Speed=2: `/src/node_modules/jxl/jxl/enc_adaptive_quantization.cc:726: JXL_ASSERT: qf_higher / qf_lower < 253`
- Speed=1: Success

In all results, transparent pixels are replaced with black. 

Using the red fox picture:

- Speed=7: Success
- Speed=6: `Cannot enlarge memory, asked to go up to 2159349760 bytes, but the limit is 2147483648 bytes!`
- Speed=5: Success
- Speed=4: Success
- Speed=3: Success
- Speed=2: `/src/node_modules/jxl/jxl/enc_adaptive_quantization.cc:726: JXL_ASSERT: qf_higher / qf_lower < 253`
- Speed=1: `jxl_enc.js:5333 Cannot enlarge memory, asked to go up to 2170884096 bytes, but the limit is 2147483648 bytes!`

It is surprising to me that we are hitting the 2GB limit, but I guess some parameters might still be wrong.

surma committed:
Fix copy/paste error

jonsneyers commented on codecs/jxl/dec/jxl_dec.cpp:
```diff
@@ -0,0 +1,73 @@
+      result[pixelOffset * 4 + 0] = clamp(red[x], 0, 255);
+      result[pixelOffset * 4 + 1] = clamp(green[x], 0, 255);
+      result[pixelOffset * 4 + 2] = clamp(blue[x], 0, 255);
+      result[pixelOffset * 4 + 3] = 255;
```
You can find the alpha channel by testing for `main->HasAlpha()` and if true `main->alpha()` gives you the alpha channel.

surma replied:
Yeah I was skipping alpha until I got the basics working. Thanks for this explanation!

surma replied:
I just added alpha support :)

jonsneyers commented on codecs/jxl/enc/jxl_enc.cpp:
```diff
@@ -0,0 +1,57 @@
+  jxl::ImageBundle* main = &io.Main();
+
+  cparams.speed_tier = static_cast<jxl::SpeedTier>(options.speed);
+  cparams.color_transform = jxl::ColorTransform::kNone;
```
Better to leave color_transform at its default.

jonsneyers replied:
Setting it to kNone is very much untested and will likely completely break things in lossy mode – this is only intended for lossless encoding, where you don't want to do irreversible color transforms. The default is kXYB which is what the lossy encoder is expecting to work with. The only reason we even allow kNone (and kYCbCr) in lossy mode is that we need it for lossless JPEG recompression.

surma replied:
Ah, so that seems to fix the assertion error at speed=2. However, now the colors are wrong. I guess I need to specify sRGB somewhere?
![Screenshot 2020-08-28 at 12 18 15](https://user-images.githubusercontent.com/234957/91555355-c7e72780-e928-11ea-9df5-64a68817aaba.png)

surma replied:
Actually, no. I use `SetFromSRGB()`, so it’s already marked as sRGB input. I think that’s why I originally ended up using `kNone` because it fixed the colors 😅

surma replied:
After adding alpha support just now, I wonder if the discolorization is a _decoder_ issue? I am doing this clamping on an `Image3F` which might be wrong? 

https://github.com/GoogleChromeLabs/squoosh/blob/321996807e7caefb8d72316d40d6e830fb2dc3c8/codecs/jxl/dec/jxl_dec.cpp#L42-L62

esx2ve replied:
Digging into the same issue here. [This issue](https://gitlab.com/wg1/jpeg-xl/-/issues/51) from the reference implementation has some more info.

jonsneyers commented on codecs/jxl/enc/jxl_enc.cpp:
```diff
@@ -0,0 +1,57 @@
+  jxl::PaddedBytes bytes;
+  jxl::ImageBundle* main = &io.Main();
+
+  cparams.speed_tier = static_cast<jxl::SpeedTier>(options.speed);
```
Maybe do `10-options.speed` so it behaves the same as in `cjpegxl`(and set the default somewhere in the middle instead of very-slow or very-fast).

jonsneyers commented on codecs/jxl/enc/jxl_enc.cpp:
```diff
@@ -0,0 +1,57 @@
+
+  cparams.speed_tier = static_cast<jxl::SpeedTier>(options.speed);
+  cparams.color_transform = jxl::ColorTransform::kNone;
+
```
You'll also want an option to set the target quality (`cparams.butteraugli_distance`) – it has a useful range from say 0.5 (near-lossless) to 8 (very lossy); the default is 1 which is one unit of just-noticeable-difference.

surma replied:
Added!

surma committed:
Expose butteraugli distance

surma committed:
Set default to 1

surma committed:
Add alpha support

RReverser committed:
Update jpeg-xl to latest upstream

RReverser committed:
Add timer to JPEG-XL

RReverser committed:
Fix encoder colors

RReverser committed:
Set default JXL encoder speed to 5

RReverser committed:
Switch decoder to C API

RReverser committed:
XYB version

veluca93 committed:
Update JXL to latest and do proper color management.

The JXL decoder now does colorspace conversion to sRGB correctly,
independently of the color space that the JPEG XL file was in.

RReverser committed:
Remove unused free_result

RReverser committed:
Simplify JxlDecoder freeing

Use unique_ptr with custom deleter to enforce consistent deallocation, no matter how we leave the function.

RReverser committed:
Use make_unique; remove unnecessary cast

RReverser committed:
Few more simplifications to JXL decoder

RReverser committed:
Fix types after rebase onto latest dev

RReverser committed:
Switch to -O3 for C++ codecs

RReverser committed:
Remove freeing todo

RReverser commented:
Ugh come on everyone has signed it...

RReverser commented:
@googlebot I consent.

RReverser committed:
Delete .gitignore

surma commented:
LGTM
